### PR TITLE
Update table rows in C

### PR DESCRIPTION
**Pull request by jeromekelleher**
This is the simplest way I could think of doing #1545

There's a lot of different ways this could be made more efficient, but I think it's probably not worth trying unless we find it's a genuine bottleneck. I had a go at doing it more efficiently and (a) it'd be a *lot* more core, and (b) there's all sorts of tricky subtleties around memcpy'ing potentially overlapping buffers and changing pointer values because of potential reallocs. This is as dumb as rocks, but should be solid.

Any thoughts @benjeffery?

### CODE REVIEW

**Comment by benjeffery** [comment]
Looks good, and pretty rock solid as it uses the existing API. I feel we're really not sure this a bottleneck so happy to proceed with this version. `TSK_NO_EDGE_METADATA` would be an easy exception if that was ever a bottleneck as that requires no reallocs.

**Comment by jeromekelleher** [comment]
OK, will tap away at this in my procrastination time.

**Comment by petrelharp** [comment]
There should probably be a comment in the docs that one should not use this method to change every row in a table if one is concerned about speed.

**Comment by jeromekelleher** [comment]
I changed the signature here to match ``add_row``, as this seemed a bit more natural as I started writing tests. Also added in a special case for the (probably common?) case in which we're replacing same-size values. This would make things like setting rows in the middle of the edge table, or say, changing the flags value for a node, much more tolerable. I think it's a nice idiom to do:
```python
tables.nodes[0] = tables.nodes[0].replace(flags=tskit.NODE_IS_SAMPLE)
```
so we should make it efficient to do this.

**Comment by jeromekelleher** [comment]
It would be good to get some eyes on this now @benjeffery before I embark on the copy-paste-athon. Notes:
- I change the name to "update_row" as this seemed clearer
- I added the special case for equal sizes, as this seems important.
- I realised that there was some nasty undefined behaviour lurking with our use of memcpy, if someone used the result of ``get_row`` as input (which seems a perfectly reasonable thing to do). This is also a possibility in the full rewrite case, so I thought it safer to change over to ``memmove`` there also. I'm sure the perf difference will be undetectable for our purposes.

Annoyingly, valgrind doesn't catch these undefined behaviours with memcpy on linux as it seems that under the hood somewhere memcpy is being replaced with memmove, and valgrind therefore doesn't see it. So, we'd have defined behaviour on some platforms, but not on others :frowning_face: 

So, my take is to change to memmove anywhere we're taking pointers as input just to be on the safe side.

**Comment by grahamgower** [comment]
> Annoyingly, valgrind doesn't catch these undefined behaviours with memcpy on linux as it seems that under the hood somewhere memcpy is being replaced with memmove, and valgrind therefore doesn't see it. So, we'd have defined behaviour on some platforms, but not on others

This is probably compiler dependent. At least with GCC, memcpy can be inlined as an ordinary copying loop (e.g. using SSE instructions).

**Review by benjeffery — COMMENTED** [review]
LGTM!
I'm happy with using `memmove` throughout - I doubt the perf gain of `memcpy` is significant.

File: c/tskit/tables.c
Comment (jeromekelleher) [reply c/tskit/tables.c:851]: Yep, that was the thinking. No point in having unreachable branches lying around, I think.

**Comment by jeromekelleher** [comment]
This is ready to go now I think - there's a lot of code, so I think it's best to wait until @benjeffery gets back to merge.

**Review by benjeffery — APPROVED** [review]
LGTM, been through closely.

**Pull request merged**
